source/common/config/config_provider_impl.h

```diff
@@ -0,0 +1,364 @@
+   * @param config supplies the newly instantiated config.
+   */
+  void onConfigUpdate(ConfigConstSharedPtr config) {
+    tls_->runOnAllThreads(
+        [this, config]() -> void { tls_->getTyped<ThreadLocalConfig>().config_ = config; });
+  }
+
+protected:
+  DynamicConfigProviderImpl(ConfigSubscriptionInstanceSharedPtr&& subscription,
+                            Server::Configuration::FactoryContext& factory_context)
+      : subscription_(subscription), tls_(factory_context.threadLocal().allocateSlot()) {}
+
+  const ConfigSubscriptionInstanceSharedPtr& subscription() const { return subscription_; }
+
+private:
+  struct ThreadLocalConfig : public ThreadLocal::ThreadLocalObject {
+    ThreadLocalConfig(ConfigProvider::ConfigConstSharedPtr initial_config)
+        : config_(initial_config) {}
+
+    ConfigProvider::ConfigConstSharedPtr config_;
+  };
+
+  ConfigSubscriptionInstanceSharedPtr subscription_;
+  ThreadLocal::SlotPtr tls_;
+};
+
+/**
+ * Provides generic functionality required by all config provider managers, such as managing shared
+ * lifetime of subscriptions and dynamic config providers, along with determining which
+ * subscriptions should be associated with newly instantiated providers.
+ *
+ * This class can not be instantiated directly; instead, it provides the foundation for
+ * dynamic config provider implementations which derive from it.
+ */
+class ConfigProviderManagerImpl : public ConfigProviderManager, public Singleton::Instance {
+public:
+  ConfigProviderManagerImpl() = delete;
+
+  virtual ~ConfigProviderManagerImpl() {}
+
+  /**
+   * This is invoked by the /config_dump admin handler.
+   * @return ProtobufTypes::MessagePtr the config dump proto corresponding to the associated
+   *                                   {static,dynamic} config providers.
+   */
+  virtual ProtobufTypes::MessagePtr dumpConfigs() const PURE;
+
+protected:
+  using ConfigProviderSet = std::unordered_set<ConfigProvider*>;
+  using ConfigSubscriptionMap =
+      std::unordered_map<std::string, std::weak_ptr<ConfigSubscriptionInstance>>;
+
+  ConfigProviderManagerImpl(Server::Admin& admin, const std::string& config_name);
+
+  const ConfigProviderSet& static_config_providers() const { return static_config_providers_; }
+
+  const ConfigSubscriptionMap& config_subscriptions() const { return config_subscriptions_; }
+
+  /**
+   * Returns the subscription associated with the config_source_proto; if none exists, a new one is
+   * allocated according to the subscription_factory_fn.
+   * @param config_source_proto supplies the proto specifying the config subscription paremeters.
+   * @param init_manager supplies the init manager.
+   * @param subscription_factory_fn supplies a function to be called when a new subscription needs
+   *                                to be allocated.
+   * @return std::shared_ptr<T> an existing (if a match is found) or newly allocated subscription.
+   */
+  template <typename T>
```

Comment:
since you already have every implementation in this class in header, I think you can move this template (and static_assert) to class level, then you don't need to worry about type consistency?

Reply:
I don't think this is easily doable. ConfigProviderManagerImpl references are being passed to several functions without concrete type awareness, including StaticConfigProvider and ConfigSubscriptionInstance constructors. This change would ripple forcing affected base classes to become templatized as well; I haven't quite fully worked it out, but unless I'm missing something, this seems like a large change with small benefit.

Reply:
That's fair, fine to leave as is.